File: C-fast/fft.c
#include <math.h>
//#include <malloc.h>
#include "fft.h"

// Internal variables
static int phasevec_exist = 0;
static double complex phasevec[32];

// Public function
void fft(int log2point, double complex *xy_out, const double complex *xy_in) {
    int i;
    if (!phasevec_exist) {
	for (i=0; i<32; i++) {
	    int point = 2<<i;
	    phasevec[i] = cos(-2*M_PI/point) + sin(-2*M_PI/point)*I;
	}
	phasevec_exist = 1;
    }
    for (i=0; i < (1<<log2point); i++) {
	unsigned int brev = i;
	brev = ((brev & 0xaaaaaaaa) >> 1) | ((brev & 0x55555555) << 1);
	brev = ((brev & 0xcccccccc) >> 2) | ((brev & 0x33333333) << 2);
	brev = ((brev & 0xf0f0f0f0) >> 4) | ((brev & 0x0f0f0f0f) << 4);
	brev = ((brev & 0xff00ff00) >> 8) | ((brev & 0x00ff00ff) << 8);
	brev = (brev >> 16) | (brev << 16);

	brev >>= 32-log2point;
	xy_out[brev] = xy_in[i];
    }

    // here begins the Danielson-Lanczos section
    int n = 1<<log2point;
    int l2pt=0;
    int mmax=1;
#ifdef MOD_SPEED
    while (l2pt < log2point) {
	int istep = 2<<l2pt;
#else
    while (n>mmax) {
	int istep = mmax<<1;
#endif
//	double theta = -2*M_PI/istep;
//	double complex wphase_XY = cos(theta) + sin(theta)*I;
	double complex wphase_XY = phasevec[l2pt++];

	double complex w_XY = 1.0 + 0.0*I;
	for (int m=0; m < mmax; m++) {
	    for (int i=m; i < n; i += istep) {
		double complex tempXY = w_XY *xy_out[i+mmax];
		xy_out[i+mmax]  = xy_out[i] - tempXY;
		xy_out[i     ] += tempXY;
	    }
	    w_XY *= wphase_XY; // rotate
	}
	mmax=istep;
    }
}

File: F#-DotNetCore/fft.fs
namespace FftPerformanceDemo

module FftModule =
    
    open System
    open System.Numerics

    let mutable phasevec_exist = false
    let mutable phasevec = Array.init 32 (fun _ -> Complex())

    for i = 0 to 32 - 1 do
        let point = 2 <<< i
        phasevec.[i] <- Complex(Math.Cos(-2.0 * Math.PI / float point), Math.Sin(-2.0 * Math.PI / float point))

    let fft log2point (xy_in: Complex array) =
        let mutable xy_out = Array.init (1 <<< log2point) (fun _ -> Complex())
    
        for i = 0 to (1 <<< log2point) - 1 do
            let mutable (brev: uint32) = uint32 i
            brev <- ((brev &&& uint32 0xaaaaaaaa) >>> 1) ||| ((brev &&& uint32 0x55555555) <<< 1)
            brev <- ((brev &&& uint32 0xcccccccc) >>> 2) ||| ((brev &&& uint32 0x33333333) <<< 2)
            brev <- ((brev &&& uint32 0xf0f0f0f0) >>> 4) ||| ((brev &&& uint32 0x0f0f0f0f) <<< 4)
            brev <- ((brev &&& uint32 0xff00ff00) >>> 8) ||| ((brev &&& uint32 0x00ff00ff) <<< 8)
            brev <- (brev >>> 16) ||| (brev <<< 16)

            brev <- brev >>> (32 - log2point)
            xy_out.[int brev] <- xy_in.[i]

        // here begins the Danielson-Lanczos section
        let n = 1 <<< log2point
        let mutable l2pt = 0
        let mutable mmax = 1

        while n > mmax do
            let istep = mmax <<< 1

            let wphase_XY = phasevec.[l2pt]
            l2pt <- l2pt + 1

            let mutable w_XY = Complex.One
    
            for m = 0 to mmax - 1 do
                let mutable i = m
                while i < n - 1 do
                    let tempXY = w_XY * xy_out.[i + mmax]
                    xy_out.[i+mmax] <- xy_out.[i] - tempXY
                    xy_out.[i] <- xy_out.[i] + tempXY

                    i <- i + istep
                w_XY <- w_XY * wphase_XY // rotate
        
            mmax <- istep
        
        xy_out